Component namespace MUST contain "\Component\"

HermanPeeren opened the pull request:
Component namespace MUST contain "\Component\", followed by the extension name. Otherwise you would get the wrong component name in ComponentHelper::getComponentName().

Pull Request for Issue #24 .

### Summary of Changes

Corrected the component namespace prefix in the text to contain  "\Component\", followed by the component name. Added a summary of the 3 parts the component namespace prefix MUST consist of, mainly in chapters "Joomla 3 MVC vs Joomla 4 MVC" and "How things work Joomla 4.0 onwards" of the components section.

Added a paragraph on the use of the the addfieldprefix attribute to tell Joomla which namespace to use for custom form fields.

### Reasoning

Although it is questionable that Joomla prescripts this format of a custom component namespace, which might not be necessary if ComponentHelper::getComponentName() would be changed, the goal is to provide a manual to build custom components that would work in the actual Joomla version.

HermanPeeren committed:
Component namespace MUST contain "\Component\"

Component namespace MUST contain "\Component\", followed by the extension name. Otherwise you would get the wrong component name in ComponentHelper::getComponentName().

nikosdion commented on sections/components.xml:
```diff
@@ -200,22 +200,52 @@ in the section entitled "GNU Free Documentation License".
-    create in your component — and there <emphasis>is</emphasis> a way to tell
-    Joomla which namespace to look custom form fields in with new attributes
-    in XML forms.</para>
+    <classname>\<replaceable>MyComponentNamespace</replaceable>\Site\Model</classname>
```
I prefer to give a more realistic example here, e.g. `<classname><replaceable>Vendor</replaceable>\<replaceable>Package</replaceable>\Component\Site\Model</classname>`. This helps “ground” the concept. You can then talk about each individual replaceable part in more length, as you are doing below.

HermanPeeren replied:
Strange the .xml~ backup file had come through, as it was in the .gitignore. Sorry, hadn't seen it.

I'll change the MyComponentNamespace to what you call a more realistic example, BUT in this comment and subsequent ones you still made a mistake with the right order: it is NOT \Acme\Example\Component, but Acme\Component\Example, so NOT \Component after the \Package name, but BEFORE it!

I used "&lt;vendor name&gt;" instead of just "vendor" to avoid confusion with literally using the word "vendor" (as is for instance used as a directory in /libraries), but maybe the angle brackets can be confusing too... will play a bit with it, using as much as possible a concrete example like \Acme\Component\Example.

nikosdion replied:
Yes, sorry, I was typing really fast because I had to leave soon to go get my kid from the school bus :)

Regarding brackets, it's not needed because we use the `<repleceable>` tag. If it's not clear I can add the brackets and/or format it in a different way with CSS (the conversion to HTML wraps it in a span with a different class). That's why I am trying to use the DocBook 5 XML features as much as possible; they may slow me down a little bit when writing the documentation but they make improving the output infinitely easier with bog standard CSS.

nikosdion commented on sections/components.xml:
```diff
@@ -200,22 +200,52 @@ in the section entitled "GNU Free Documentation License".
+      </listitem>
+    </itemizedlist>
+
+    <para>The namespace of your custom component is also used for custom form
```
Technically, it can be used for many different things. Can you please rephrase to something along the lines of “The namespace of your component can also be used elsewhere where a class which belongs to your component might be used. For example, in custom form fields you may…”

nikosdion commented on sections/components.xml:
```diff
@@ -200,22 +200,52 @@ in the section entitled "GNU Free Documentation License".
+    fields you may need to create in your component. The way to tell Joomla
+    which namespace to use for custom form fields is with the
+    <code>addfieldprefix</code> attribute in an XML form. This attribute of a
+    fieldset (and method of Joomla\CMS\Form\FormHelper) is new since Joomla
```
Please wrap the class name with `<classname>` for consistency. I know I may have missed a few spots myself, I try to fix them as I re-read what I've already written ;)

nikosdion commented on sections/components.xml:
```diff
@@ -399,9 +429,9 @@ in the section entitled "GNU Free Documentation License".
-      <code>\My\Prefix\Dispatcher\Dispatcher</code>, where
-      <code>\My\Prefix</code> is your component's prefix. If this class
-      exists, an object instance of it is returned.</para>
+      <code>\MyComponentNamespace\Dispatcher\Dispatcher</code>, where
```
Let's replace that here, too, with a more realistic classname.

I noticed that 3PDs reading this book copy from it (as intended!), replacing namespaces with their owns. Keeping things consistent helps them with that.

nikosdion commented on sections/components.xml:
```diff
@@ -1256,16 +1287,16 @@ return new class implements ServiceProviderInterface {
-      <code><replaceable>\My\Prefix</replaceable>\Administrator\Extension\<replaceable>Name</replaceable>Component</code>
-      where \My\Prefix is the namespace prefix of your component and
-      <code>Name</code> is the name of your component without the
+      <code><replaceable>\MyComponentNamespace</replaceable>\Administrator\Extension\<replaceable>Name</replaceable>Component</code>
```
Same thing here, let's go for a more realistic namespace.

nikosdion commented on sections/components.xml:
```diff
@@ -1407,13 +1438,13 @@ public static function getContainer()
     <para>If you do decide to create a custom Dispatcher its class name
     <emphasis role="bold">MUST</emphasis> be
-    <code><replaceable>\My\Prefix</replaceable>\Administrator\Dispatcher\Dispatcher</code>
+    <code><replaceable>\MyComponentNamespace</replaceable>\Administrator\Dispatcher\Dispatcher</code>
```
Use a realistic namespace.

nikosdion commented on sections/components.xml:
```diff
@@ -1626,58 +1657,80 @@ class Dispatcher extends \Joomla\CMS\Dispatcher\ComponentDispatcher
       <code>&lt;extension&gt;</code> root element:</para>
 
-      <programlisting language="xml">&lt;namespace path="src"&gt;My\Namespace\Prefix&lt;/namespace&gt;</programlisting>
+      <programlisting language="xml">&lt;namespace path="src"&gt;MyComponentNamespace&lt;/namespace&gt;</programlisting>
```
Use a realistic namespace.

nikosdion commented on sections/components.xml:
```diff
@@ -1626,58 +1657,80 @@ class Dispatcher extends \Joomla\CMS\Dispatcher\ComponentDispatcher
       <filename>src</filename>.</para>
 
-      <para>The text inside the XML element, <code>My\Namespace\Prefix</code>
+      <para>The text inside the XML element, <code>MyComponentNamespace</code>
```
Use a realistic namespace.

nikosdion commented on sections/components.xml:
```diff
@@ -1626,58 +1657,80 @@ class Dispatcher extends \Joomla\CMS\Dispatcher\ComponentDispatcher
+          unique to you, your company or your project. It should not start
+          with <code>\Joomla</code>, because component namespaces starting
+          withIn <code>\Joomla</code> are reserved for core components. And it
+          should not contain <code>\Component</code>, as that would interfere
```
Confusing syntax. Best clarify as “Component should only appear as the last part of the namespace prefix, e.g. Foo\Bar\Component but NOT Foo\Component\Bar\Component. The word Component has a special meaning in Joomla and should best be avoided in the customisable parts of your namespace prefix”. Giving a good example like you do later one is great!

HermanPeeren replied:
No: the component name must be namespace part immediately after the FIRST occurrence of "\Component". 

So:
* NOT: “Component should only appear as the last part of the namespace prefix, e.g. Foo\Bar\Component"
* but "Foo\Component\Bar\Component", although not advisable, would not be a problem.

I'll try to make it as clear as possible.

nikosdion reviewed (changes requested):
Thank you, Herman! I saw your issue yesterday but I never had the chance to rewrite those chapters. Thank you for doing the work :)

I have a few, relatively minor requests to make things more clear to the intended target audience, based on the feedback I have received across different media (unfortunately, only part of it is in the GitHub issues).

Also, could you please remove the `.xml~` backup file from the PR? I think it was added on accident.

Thank you in advance!

HermanPeeren committed:
Delete components.xml~

HermanPeeren committed:
Use a more realistic namespace

Some minor editing based on Nic's review 15-12-2022: Use a more realistic abstract namespace, use replaceable tag instead of angle brackets, clarify not using \Component sub-namespace in vendor namespace prefix. Also added a basic explanation of PSR4 under "Namespaces" in "Joomla 3 MVC vs Joomla 4 MVC".

HermanPeeren commented:
Requested changes done.

nikosdion commented:
Thank you very much!
